### PR TITLE
Convert MessageQueue to TypeScript

### PR DESCRIPTION
**Pull request by owenpearson**

### CODE REVIEW

**Review by QuintinWillison — APPROVED** [review]

File: common/lib/transport/messagequeue.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	shift(): PendingMessage | undefined {
+		return this.messages.shift();
+	}
+
+	last(): PendingMessage {
+		return this.messages[this.messages.length - 1];
+	}
+
+	copyAll(): PendingMessage[] {
+		return this.messages.slice();
```
[the remark this answers is not in the document]
Comment (owenpearson) [reply common/lib/transport/messagequeue.ts:32]: I'm guessing you mean as opposed to a deep copy? The messages in the queue aren't mutated so the only purpose of this function is to make sure messages aren't added/removed from the result after it's called, in which case a shallow copy works fine.

File: common/lib/transport/messagequeue.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+	prepend(messages: Array<PendingMessage>): void {
+		this.messages.unshift.apply(this.messages, messages);
+	}
+
+	completeMessages(serial: number, count: number, err?: ErrorInfo | null): void {
+		Logger.logAction(Logger.LOG_MICRO, 'MessageQueue.completeMessages()', 'serial = ' + serial + '; count = ' + count);
+		err = err || null;
+		const messages = this.messages;
+		const first = messages[0];
```
[the remark this answers is not in the document]
Comment (owenpearson) [reply common/lib/transport/messagequeue.ts:53]: Thanks, fixed in 15e20a5

**Review by ikbalkaya — APPROVED** [review]
LGTM

File: common/lib/transport/messagequeue.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	push(message: PendingMessage): void {
+		this.messages.push(message);
+	}
+
+	shift(): PendingMessage | undefined {
+		return this.messages.shift();
+	}
+
+	last(): PendingMessage {
+		return this.messages[this.messages.length - 1];
```
[the remark this answers is not in the document]
Comment (owenpearson) [reply common/lib/transport/messagequeue.ts:28]: Ah, I was wrong. If `this.messages.length === 0` then this method simply returns `undefined`. Throwing an error in that case actually broke some of the tests 😄

**Pull request merged**